public_html/wp-content/themes/pattern-directory/src/components/pattern-thumbnail/screenshot.js

```diff
@@ -0,0 +1,117 @@
+/* global FileReader, fetch */
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { addQueryArgs } from '@wordpress/url';
+import { Spinner } from '@wordpress/components';
+import { useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useInterval from '../../hooks/use-interval';
+
+/**
+ * Module constants
+ */
+const MAX_ATTEMPTS = 10;
+const RETRY_DELAY = 1000;
+
+/**
+ *
+ * We are using mShots which is a project that generates a screenshot from a webpage.
+ * Since the screenshot generation takes some time, for never seen websites,
+ * we need to do some custom handling.
+ *
+ * @param {Object}  props
+ * @param {string}  props.alt       The alt text for the screenshot image.
+ * @param {string}  props.className Any extra class names for the wrapper div.
+ * @param {boolean} props.isReady   Whether we should start try to show the image.
+ * @param {Object}  props.size      The viewport size to take the screenshot at.
+ * @param {Object}  props.style     Styles for the wrapper div.
+ * @param {string}  props.src       The url of the page to screenshot.
+ * @return {Object} React component
+ */
+export default function ( { alt, className, isReady = false, size, src, style } ) {
+	const fullUrl = addQueryArgs( `https://s0.wp.com/mshots/v1/${ encodeURIComponent( src ) }`, {
+		vpw: size.width,
+		vph: size.height,
+	} );
+
+	const [ attempts, setAttempts ] = useState( 0 );
+	const [ hasLoaded, setHasLoaded ] = useState( false );
+	const [ hasError, setHasError ] = useState( false );
+	const [ base64Img, setBase64Img ] = useState( '' );
+
+	// We don't want to keep trying infinitely.
+	const hasAborted = attempts > MAX_ATTEMPTS;
+
+	// The derived loading state
+	const isLoading = isReady && ! hasLoaded && ! hasAborted && ! hasError;
+
+	/**
+	 * Since we already made the request, we'll use the response to be frugal.
+	 *
+	 * @param {string} res
+	 */
+	const convertResponseToBase64 = async ( res ) => {
+		const blob = await res.blob();
+
+		const reader = new FileReader();
+		reader.onload = ( event ) => {
+			setBase64Img( event.target.result );
+		};
+		reader.readAsDataURL( blob );
+	};
+
+	/**
+	 * The Snapshot service will redirect when its generating an image.
+	 * We want to continue requesting the image until it doesn't redirect.
+	 */
+	useInterval(
+		async () => {
+			try {
+				const res = await fetch( fullUrl );
+
+				if ( res.status === 200 && ! res.redirected ) {
+					await convertResponseToBase64( res );
+
+					setHasLoaded( true );
+				} else {
+					setAttempts( attempts + 1 );
+				}
+			} catch ( error ) {
+				setHasError( true );
+			}
+		},
+		isLoading ? RETRY_DELAY : null
+	);
+
+	if ( ! isReady ) {
+		return null;
+	}
+
+	if ( isLoading ) {
+		return (
+			<div className={ `${ className } is-loading` } style={ style }>
+				<Spinner style={ { width: '33%', height: '33%' } } />
```

Comment:
The spinner feels a little bit large, but that's personal preference.

Reply:
Fair enough - the default size was very tiny, and this wasn't really chosen with any intention. Swapped it out for 2x default size.

<img width="336" alt="Screen Shot 2022-10-06 at 12 27 49 PM" src="https://user-images.githubusercontent.com/541093/194367815-7bfc6752-eb5b-46ea-9adf-066c50adf623.png">